src/normal.c

```diff
@@ -346,6 +347,72 @@ static const struct nv_cmd
     {K_PS, nv_edit, 0, 0},
 };
 
+#define strstartswith(a, b) (!strncmp(a, b, strlen(b)))
+
+void toggle_comment(linenr_T lnum)
+{
+  const char_u *comment = curbuf->b_oni_line_comment != NULL ? curbuf->b_oni_line_comment : (char_u *)"//";
+  int commentlen = (int)STRLEN(comment);
+  const char_u *line = ml_get(lnum);
+  int linelen = (int)STRLEN(line);
+  char_u *newp;
+
+  if (strstartswith(line, comment))
+  {
+    // remove comment
+
+    newp = alloc((linelen - commentlen) + 1);
+
+    if (newp == NULL)
+      return;
+
+    if (virtual_active() && curwin->w_cursor.coladd > 0)
+      coladvance_force(getviscol());
+
+    mch_memmove(newp, line + commentlen, (size_t)((linelen - commentlen) + 1));
+    ml_replace(lnum, newp, FALSE);
+  }
+  else
+  {
+    // add comment
+
+    newp = alloc(linelen + commentlen + 1);
+
+    if (newp == NULL)
+      return;
+
+    if (virtual_active() && curwin->w_cursor.coladd > 0)
+      coladvance_force(getviscol());
+
+    mch_memmove(newp, comment, (size_t)commentlen);
+    mch_memmove(newp + commentlen, line, (size_t)(linelen + 1));
+    ml_replace(lnum, newp, FALSE);
+    inserted_bytes(lnum, 0, commentlen);
+  }
+}
+
+void toggle_comment_lines(linenr_T start, linenr_T end)
+{
+  linenr_T lnum;
```

Comment:
It doesn't look like undo is hooked up, testing in the UI. I think we need a `u_save` call from `start - 1` to `end + 1` - example here: https://github.com/onivim/libvim/blob/45b7a7399488792889bd1ccc1ec98825ed116bb5/src/ops.c#L200

Reply:
Yep, that did the trick! Strange that the undo tests still pass without it though.